src/show.jl

```diff
@@ -126,85 +119,107 @@ Configurable option to control whether emoji icons (`true`) or a plain-text anno
 """
 const SHOW_TREE_ICONS = Ref{Bool}(true)
 
+"""
+    SHOW_TREE_MAX_DEPTH = Ref{Int}(5)
+
+Maximum recursive depth to descend during printing.
+"""
+const SHOW_TREE_MAX_DEPTH = Ref{Int}(5)
+
+"""
+    SHOW_TREE_MAX_CHILDREN = Ref{Int}(50)
+
+Maximum number of children to show at each node.
+"""
+const SHOW_TREE_MAX_CHILDREN = Ref{Int}(50)
+
 function Base.show(io::IO, ::MIME"text/plain", obj::Union{File,Group,Dataset,Attributes,Attribute})
-    if SHOW_TREE[]
-        show_tree(io, obj)
-    else
+    if get(io, :compact, false)::Bool
         show(io, obj)
-    end
-end
-
-function _tree_icon(obj)
-    if SHOW_TREE_ICONS[]
-        return obj isa Attribute ? "🏷️" :
-               obj isa Group ? "📂" :
-               obj isa Dataset ? "🔢" :
-               obj isa Datatype ? "📄" :
-               obj isa File ? "🗂️" :
-               "❓"
     else
-        return obj isa Attribute ? "[A]" :
-               obj isa Group ? "[G]" :
-               obj isa Dataset ? "[D]" :
-               obj isa Datatype ? "[T]" :
-               obj isa File ? "[F]" :
-               "[?]"
+        show_tree(io, obj)
     end
 end
+
+_tree_icon(::Type{Attribute}) = SHOW_TREE_ICONS[] ? "🏷️" : "[A]"
+_tree_icon(::Type{Group})     = SHOW_TREE_ICONS[] ? "📂" : "[G]"
+_tree_icon(::Type{Dataset})   = SHOW_TREE_ICONS[] ? "🔢" : "[D]"
+_tree_icon(::Type{Datatype})  = SHOW_TREE_ICONS[] ? "📄" : "[T]"
+_tree_icon(::Type{File})      = SHOW_TREE_ICONS[] ? "🗂️" : "[F]"
+_tree_icon(::Type)            = SHOW_TREE_ICONS[] ? "❓" : "[?]"
+_tree_icon(obj) = _tree_icon(typeof(obj))
 _tree_icon(obj::Attributes) = _tree_icon(obj.parent)
 
 _tree_head(io::IO, obj) = print(io, _tree_icon(obj), " ", obj)
 _tree_head(io::IO, obj::Datatype) = print(io, _tree_icon(obj), " HDF5.Datatype: ", name(obj))
 
-function _tree_children(parent::Union{File, Group}, attributes::Bool)
-    names = keys(parent)
-    objs  = Union{Object, Attribute}[parent[n] for n in names]
-    if attributes
-        attrn = keys(HDF5.attributes(parent))
-        attro = Union{Object, Attribute}[HDF5.attributes(parent)[n] for n in attrn]
-        names = append!(attrn, names)
-        objs  = append!(attro, objs)
+_tree_count(parent::Union{File,Group}, attributes::Bool) =
+    length(parent) + (attributes ? length(HDF5.attributes(parent)) : 0)
+_tree_count(parent::Dataset, attributes::Bool) =
+    attributes ? length(HDF5.attributes(parent)) : 0
+_tree_count(parent::Attributes, _::Bool) = length(parent)
+_tree_count(parent::Union{Attribute,Datatype}, _::Bool) = 0
+
+function _show_tree(io::IO, obj::Union{File,Group,Dataset,Datatype,Attributes,Attribute}, indent::String="";
+                    attributes::Bool = true, depth::Int = 1)
+    isempty(indent) && _tree_head(io, obj)
```

Comment:
I feel like we should instead use `displaysize(io)` to determine a suitable truncation point instead of the heuristics.

Reply:
I agree in spirit, though it'll probably take another revamp of this PR to do that. I think more contextual information will have to be sent through the recursions — let me mull that over.

Reply:
Yeah the recursive tree printing makes this  more complicated.  Feel free to ignore this comment for now.

Reply:
We can open an issue to make sure this goal isn't forgotten.